crypto/rsa/rsa_lib.c

```diff
@@ -897,6 +897,63 @@ int EVP_PKEY_CTX_get_rsa_padding(EVP_PKEY_CTX *ctx, int *pad_mode)
 
 }
 
+int EVP_PKEY_CTX_set_rsa_pss_keygen_md(EVP_PKEY_CTX *ctx, const EVP_MD *md)
+{
+    const char *name;
+
+    if (ctx == NULL || md == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
+
+    /* If key type not RSA return error */
+    if (ctx->pmeth != NULL && ctx->pmeth->pkey_id != EVP_PKEY_RSA &&
+        ctx->pmeth->pkey_id != EVP_PKEY_RSA_PSS)
+        return -1;
+
+    name = (md == NULL) ? "" : EVP_MD_name(md);
```

Comment:
Ok, it's looking better.  However, we do need legacy support *somewhere* (because engines), and this would be the spot, as far as I can tell.  Here's the code that I would insert, right here:

``` C
    /* TODO(3.0): Remove this eventually when no more legacy */
    if (ctx->op.keymgmt.genctx == NULL)
        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_RSA_PSS, EVP_PKEY_OP_KEYGEN,
                                 EVP_PKEY_CTRL_MD, 0, (void *)md);
```

Reply:
Er, I would insert that *before* the `name` assignment.

Reply:
Thanks, i was not sure how to handle the legacy case. I've update now.